### ISSUE TITLE
require can be very slow with lot of gems

### ISSUE THREAD

**Issue by mfazekas** [rubygems/rubygems#1167]
I have a state of rubygems installed where i cannot start up the rails command without a Gemfile - like rails new foo . Debugging showed the it's not infinite loop just very very slow in in ```find_in_unresolved_tree```. I have a lot of version of lot of gems installed.

the following testcase should demonstrate the problem.  With ruby 2.1.1 and master of rubygems the code bellow - emulating a require with 7 gems and each with 3 versions with relaxed version requirements takes 26 seconds.

Some memoization/dynamic programming could help in speeding this up - as ```find_in_unresolved_tree``` might be exponentiental in the number of gems (if each gem has multiple compatible versions) and the file is not found quickly.

```ruby
  def test_find_in_unresolved_tree
    begin
      num_of_pkg = 7
      num_of_version_per_pkg = 3
      packages = (0..num_of_pkg).map do |pkgi|
        (0..num_of_version_per_pkg).map do |pkg_version|
          deps = Hash[(pkgi..num_of_pkg).map { |deppkgi| ["pkg#{deppkgi}", ">= 0"] }]
          rsepc = new_spec "pkg#{pkgi}", pkg_version.to_s, deps
          Gem::Specification.add_spec rsepc
          rsepc
        end
      end
      unresolved = Gem::Specification.unresolved_deps
    
      packages.flatten.each do |pkg|
        pkg.dependencies.each do |dep|
          unresolved[dep.name].merge dep
        end
      end

      specs = unresolved.values.map { |dep| dep.to_specs }.flatten
      require 'benchmark'
      puts Benchmark.measure { 
        Gem::Specification.find_in_unresolved_tree 'foo'
      }
    ensure
      Gem::Specification.unresolved_deps.clear
      Gem::Specification.reset
    end
  end
```

**Comment by tenderlove** [comment]
Oh great!  Thank you for this repro script.  I will see what I can do about speeding it up!!  In which file did you add this test case?

**Comment by mfazekas** [comment]
into /test/rubygems/test_gem_specification.rb

**Comment by mfazekas** [comment]
Added a potential fix for performance: #1168 
But logic is also broken #1169 and performance fix does not correct that.

**Comment by schneems** [comment]
@mfazekas thanks for reporting this! I'm curious about what techniques and tools you used to find the issue was in `find_in_unresolved_tree `

**Comment by mfazekas** [comment]
@schneems I've got into a state where my ```rails new appname``` command seemd to take forever. I've used dtrace to find out what's which ruby methods were executing. Then i've used something like ```byebug if $itercount==10000``` in one of those frequently called method to understand where and why they are called from that many times.

```sh
sudo dtrace -n \
  'ruby*:::method-entry { @[copyinstr(arg0), copyinstr(arg1)] = count() }' \
  -p <pid> 
```

**Comment by lzap** [comment]
Found this randomly, just for other visitors, similar tool that is built-in Fedora and RHEL is SystemTap: http://lukas.zapletalovi.com/2012/02/peek-into-your-ruby-app-with-systemtap.html

I have been investigating the very same issue, but on the filesystem level some time ago. Nice catch!

**Comment by copiousfreetime** [comment]
This appears to have been resolved with #1188
